File: src/ir/module.cpp
#include "toyc/ir/module.h"

void Module::add_function(std::unique_ptr<Function> fn) {
    if (fn->name() == "main") {
        main_ = fn.get();
    }
    functions_.push_back(std::move(fn));
}

void Module::add_global(GlobalVar* gv) {
    globals_.push_back(gv);
}

void Module::add_global_constant(Constant* c) {
    global_constants_.push_back(c);
}

GlobalVar* Module::new_global_var(const std::string& name, int32_t initial_value) {
    auto gv = std::make_unique<GlobalVar>(name, initial_value);
    auto* raw = gv.get();
    owned_values_.push_back(std::move(gv));
    globals_.push_back(raw);
    return raw;
}

Constant* Module::new_global_constant(const std::string& name, int32_t value) {
    auto c = std::unique_ptr<Constant>(Constant::named(next_global_const_id_++, name, value));
    auto* raw = c.get();
    owned_values_.push_back(std::move(c));
    global_constants_.push_back(raw);
    return raw;
}

Constant* Module::new_constant(int32_t value) {
    auto c = std::unique_ptr<Constant>(Constant::of(next_global_const_id_++, value));
    auto* raw = c.get();
    owned_values_.push_back(std::move(c));
    return raw;
}

File: toy-c-compiler-master/src/main/java/toyc/ir/block/Module.java
package toyc.ir.block;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import toyc.ir.IRVisitor;
import toyc.ir.value.Constant;
import toyc.ir.value.GlobalVar;

public final class Module {
    private final List<GlobalVar> globals = new ArrayList<>();
    private final List<Constant> globalConstants = new ArrayList<>();
    private final List<Function> functions = new ArrayList<>();
    private Function mainFunction;

    public List<GlobalVar> globals() {
        return Collections.unmodifiableList(globals);
    }

    public List<Constant> globalConstants() {
        return Collections.unmodifiableList(globalConstants);
    }

    public List<Function> functions() {
        return Collections.unmodifiableList(functions);
    }

    public Function mainFunction() {
        return mainFunction;
    }

    public void addGlobal(GlobalVar global) {
        globals.add(global);
    }

    public void addGlobalConstant(Constant constant) {
        globalConstants.add(constant);
    }

    public void addFunction(Function function) {
        functions.add(function);
        if ("main".equals(function.name())) {
            mainFunction = function;
        }
    }

    public void replaceFunctions(List<Function> newFunctions) {
        functions.clear();
        mainFunction = null;
        for (Function function : newFunctions) {
            addFunction(function);
        }
    }

    public <R, C> R accept(IRVisitor<R, C> visitor, C context) {
        return visitor.visitModule(this, context);
    }
}
